---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: true
    toc_float: true
    css: styles.css
---

<div class="contentbox">

## Summary

Your project description

Sample sheet used for analysis:

```{r Initialise, echo=F,result='hide',error=F,warning=F,message=F}
require(rmarkdown)
require(knitr)
require(Cairo)
knitr::opts_chunk[['set']](dev='CairoPNG')
source('ChIP.shinyrep.helpers.R')
loadGlobalVars()
isPaired <- tolower(SHINYREPS_PAIRED) == "yes"
runCutadapt <- SHINYREPS_RUN_CUTADAPT=="true"
runPeakAnno <- SHINYREPS_RUN_PEAK_ANNOTATION=="true"
runDiffbind <- SHINYREPS_RUN_DIFFBIND=="true"
runEnrich <- SHINYREPS_RUN_ENRICHMENT=="true"
SHINYREPS_FASTQC_SUMMARIZED <- as.logical(SHINYREPS_FASTQC_SUMMARIZED)

targets <<- ChIPhelper.init("readTargets")

## parallel pipeline branches
# most modules store results of different pipeline branches in separated sub-directories of the respective output folder. 
# One sub-directory may be "" as default output folder.
# Modules producing output files with names uniquely distinguishable for all branches may be stored in the same output folder. 
# If a subset of these files shall be selected they can be grepped by a character substring
branches <- c(filtered="filtered", unfiltered="unfiltered") # value is sub-directory, name is file tag in case of subsetting  

# print targets file
DT::datatable(targets, caption="targets file (sample sheet) used for analysis")

``` 


## Raw reads QC {.tabset .tabset-pills}

The raw sequence reads of all samples are analyzed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).


```{r FastQC_paragraph, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}
# Explanatory text to be displayed alongside the selected FastQC plots.
qctext <- c(
  Summary="\n\n### Summary\n\n Inspecting the PASS/WARN/FAIL Status of each FastQC module.\n\n",
  BaseQuals="\n\n### Per base sequence quality\n\n The 'Read qualities' plot summarizes the mean quality score at each position. The color on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (yellow), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area. Due to binning of quality score in recent sequencing technologies (e.g. Illumina NextSeq), sequencing qualities are the same on many position and, thus, also samples appear to have very similar or almost identical sequencing qualities.\n\n",
  SeqQuals="\n\n### Per sequence quality scores\n\n Mean Sequence Quality Per Read report allows you to see if a subset of your sequences have universally low quality values. It is often the case that a subset of sequences will have universally poor quality, often because they are poorly imaged (on the edge of the field of view etc), however these should represent only a small percentage of the total sequences.\n\n",
  SeqContent="\n\n### Per base sequence content\n\n The 'Sequence bias' plots show the proportion of each base (% G, A, T and C) at each position. In a typical ChIP-seq library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other, G being similar to C and A to T.\n\n",
  GcContent="\n\n### Per sequence GC content\n\n The 'GC content' plot shows the GC% distribution of all reads and all samples. Typically, the lines should tightly overlap and look essentially the same in all samples. An unusually shaped distribution could indicate a contaminated library.\n\n",
  DupLevels="\n\n### Sequence Duplication Levels\n\n The 'Duplication levels' plot shows the proportion of the library which is made up of sequences in each of the different duplication level bins. It is useful for the assessment of the NGS library complexity and the presence or absence of PCR over-amplification during library prep. There are two lines on the plot. The blue line takes the full sequence set and shows how its duplication levels are distributed. In the red plot the sequences are de-duplicated and the proportions shown are the proportions of the deduplicated set which come from different duplication levels in the original data. In a properly diverse library most sequences should fall into the far left of the plot in both the red and blue lines. A general level of enrichment, indicating broad oversequencing in the library will tend to flatten the lines, lowering the low end and generally raising other categories. The presence of dominant sequences (e.g. adapter dimers) will tend to produce spikes towards the right of the plot. These high duplication peaks will most often appear in the red trace as they make up a high proportion of the original library, but usually disappear in the blue trace as they make up an insignificant proportion of the deduplicated set. If peaks persist in the blue trace then this suggests that there are a large number of different highly duplicated sequences which might indicate either a contaminant set or a severe PCR duplication. See the related PBC metrics below.\n\n", 
  Overrep="\n\n### Overrepresented sequences\n\n Overrepresented Sequences lists all of the sequence which make up more than 0.1% of the total. A normal high-throughput library will contain a diverse set of sequences, with no individual sequence making up a tiny fraction of the whole. Finding that a single sequence is very overrepresented in the set either means that it is highly biologically significant, or indicates that the library is contaminated, or not as diverse as you expected.\n\n",
  AdapterContent="\n\n### Adapter Content\n\n Adapter Content does a generic analysis of all of the Kmers in your library to find those which do not have even coverage through the length of your reads. This can find a number of different sources of bias in the library which can include the presence of read-through adapter sequences building up on the end of your sequences.\n\n"
)

## If you prefer heatmaps instead of the line plots given in the chunks below, un-comment the following helper function
# qclist <- ChIPhelper.ngsReports.Fastqc(metrics=c("BaseQuals", "SeqContent", "GcContent", "DupLevels"))
# for (qc in names(qclist)) {
#   cat(qctext[qc])
#   print(qclist[[qc]])
# }

```



```{r FastQC_qual_summarized, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}
## To display individual FastQC plots define the 'metrics' parameter with any combination of the following plot names:
## "Summary", "BaseQuals", "SeqQuals", "SeqContent", "GcContent", "DupLevels", "Overrep", "AdapterContent"
## Remark: Due to different figure size requirements, some of the FastQC plots are entered in separate code chunks.
fastqc_summary_plot <- ChIPhelper.Fastqc.custom(web=F,summarizedPlots=SHINYREPS_FASTQC_SUMMARIZED,
                                                metrics=c("BaseQuals", "SeqContent", "GcContent", "DupLevels"))

if("Summary" %in% names(fastqc_summary_plot)) {
  cat(qctext["Summary"])
  print(fastqc_summary_plot[["Summary"]])
}

if (SHINYREPS_FASTQC_SUMMARIZED && "BaseQuals" %in% names(fastqc_summary_plot)) {
    cat(qctext["BaseQuals"])
    plot(fastqc_summary_plot[["BaseQuals"]])
}
```

```{r FastQC_qual, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1}
if (!SHINYREPS_FASTQC_SUMMARIZED && "BaseQuals" %in% names(fastqc_summary_plot)) {
    cat(qctext["BaseQuals"])
    plot(fastqc_summary_plot[["BaseQuals"]])
}
```

```{r FastQC_nuc_content_summarized, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', fig.height=min(170, ceiling(0.12*fastqc_summary_plot$no.of.samples)+1)}
if (SHINYREPS_FASTQC_SUMMARIZED && "SeqContent" %in% names(fastqc_summary_plot)) {
    cat(qctext["SeqContent"])
    plot(fastqc_summary_plot[["SeqContent"]])
} 
```

```{r FastQC_nuc_content, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1}
if (!SHINYREPS_FASTQC_SUMMARIZED && "SeqContent" %in% names(fastqc_summary_plot)) {
    cat(qctext["SeqContent"])
    plot(fastqc_summary_plot[["SeqContent"]])
} 
```

```{r FastQC_further_plots, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}
## Remark: knitr doesn't like plotly plots in loop structures

if("SeqQuals" %in% names(fastqc_summary_plot)) {
  cat(qctext["SeqQuals"])
  fastqc_summary_plot[["SeqQuals"]]
}
if("GcContent" %in% names(fastqc_summary_plot)) {
  cat(qctext["GcContent"])
  fastqc_summary_plot[["GcContent"]]
}
if("DupLevels" %in% names(fastqc_summary_plot)) {
  cat(qctext["DupLevels"])
  fastqc_summary_plot[["DupLevels"]]
}
if("Overrep" %in% names(fastqc_summary_plot)) {
  cat(qctext["Overrep"])
  fastqc_summary_plot[["Overrep"]]
}
if("AdapterContent" %in% names(fastqc_summary_plot)) {
  cat(qctext["AdapterContent"])
  fastqc_summary_plot[["AdapterContent"]]
}
```

## {-}


```{r cutadapt, echo=F, results='asis', error=F, warning=F, message=F, eval=runCutadapt}
##### parameters to set:
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
# The function will try to map the data from targets$sample to the cutadapt log file names (for this, the unique part given in  
# sampleColumnName must be a substring of file the name). Otherwise pruned file names will be used by default. 
# In case of pooled library design the targets object will be modified accordingly. 
#####

cat("## Adapter trimming with Cutadapt

The following plot shows the amount of reads trimmed for the selected adapter sequences. The column 'tooshort' gives the percentage of reads removed due to a length cutoff after trimming. Additional to the column 'trimmed', which is for all adapters combined, there are columns for every adapter giving the percentage of reads trimmed for this adapter.\n")

ChIPhelper.cutadapt(targetsdf=targets, colorByFactor="group")
```


```{r FastQC_qual_summarized_cutadapt, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', eval=runCutadapt}
## To display individual FastQC plots define the 'metrics' parameter with any combination of the following plot names:
## "Summary", "BaseQuals", "SeqQuals", "SeqContent", "GcContent", "DupLevels", "Overrep", "AdapterContent"

cat("### FastQC after adapter trimming {.tabset .tabset-pills}\n")

fastqc_summary_plot_cutadapt <- ChIPhelper.Fastqc.custom(web=F,subdir="trimmed", summarizedPlots=SHINYREPS_FASTQC_SUMMARIZED,
                                                metrics=c("BaseQuals", "SeqContent", "GcContent"))

if("Summary" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["Summary"])
  print(fastqc_summary_plot_cutadapt[["Summary"]])
}

if (SHINYREPS_FASTQC_SUMMARIZED && "BaseQuals" %in% names(fastqc_summary_plot_cutadapt)) {
    cat(qctext["BaseQuals"])
    plot(fastqc_summary_plot_cutadapt[["BaseQuals"]])
}
```

```{r FastQC_qual_cutadapt, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1, eval=runCutadapt}
if (SHINYREPS_FASTQC_SUMMARIZED == FALSE && "BaseQuals" %in% names(fastqc_summary_plot_cutadapt)) {
    cat(qctext["BaseQuals"])
    plot(fastqc_summary_plot_cutadapt[["BaseQuals"]])
}
```

```{r FastQC_nuc_content_summarized_cutadapt, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', fig.height=min(170, ceiling(0.12*fastqc_summary_plot$no.of.samples)+1), eval=runCutadapt}
if (SHINYREPS_FASTQC_SUMMARIZED && "SeqContent" %in% names(fastqc_summary_plot_cutadapt)) {
    cat(qctext["SeqContent"])
    plot(fastqc_summary_plot_cutadapt[["SeqContent"]])
} 
```

```{r FastQC_nuc_content_cutadapt, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1, eval=runCutadapt}
if (SHINYREPS_FASTQC_SUMMARIZED == FALSE && "SeqContent" %in% names(fastqc_summary_plot_cutadapt)) {
    cat(qctext["SeqContent"])
    plot(fastqc_summary_plot_cutadapt[["SeqContent"]])
} 
```

```{r FastQC_further_plots_cutadapt, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', eval=runCutadapt}
## Remark: knitr doesn't like plotly plots in loop structures

if("SeqQuals" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["SeqQuals"])
  fastqc_summary_plot_cutadapt[["SeqQuals"]]
}
if("GcContent" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["GcContent"])
  fastqc_summary_plot_cutadapt[["GcContent"]]
}
if("DupLevels" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["DupLevels"])
  fastqc_summary_plot_cutadapt[["DupLevels"]]
}
if("Overrep" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["Overrep"])
  fastqc_summary_plot_cutadapt[["Overrep"]]
}
if("AdapterContent" %in% names(fastqc_summary_plot_cutadapt)) {
  cat(qctext["AdapterContent"])
  fastqc_summary_plot_cutadapt[["AdapterContent"]]
}


cat("\n", fill=T)
cat("### {-}\n", fill = T)

```

## Read mapping ##

Mapping to the reference genome was done using the established NGS mapping tool Bowtie2 (http://bowtie-bio.sourceforge.net/bowtie2). 

```{r Bowtie2_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
# mapping statistics for either paired end or single read design

if(isPaired) {cat(" The mapping statistics show the total number of reads delivered to the aligner ('all pairs'), the number of read mapping concordantly to one location ('unique'), the number of read pairs mapping concordantly to multiple locations ('multi'), the number of read pairs mapping discordantly to one location ('discordantly'), the number of reads pairs where only one reads maps to one location ('single unique'), the number of read pairs where only one read maps to multiple locations ('single multi') and the overall alignment rate (this is based on the amount of pairs aligned, read pairs in which only one read is mapped are taken into account as half a pair). Percentages are based on the amount of pairs sequenced ('all pairs').

A read pair that aligns with the expected relative mate orientation (typically 'fr') and with the expected range of distances between mates (typically up to 1 kb) is said to align 'concordantly'. If both mates have unique alignments, but the alignments do not match paired-end expectations, the pair is said to align 'discordantly'. If Bowtie2 cannot find a paired-end alignment for a pair, by default it will search for unpaired alignments of the respective read mates. For most ChIP-seq experiments, discordant and multi-mapping reads are typically discarded in downstream analyses.\n\n")
  
  } else {
    
    cat(" The mapping statistics show the total number of reads delivered to the aligner ('all reads'), the number of reads mapping uniquely to one location ('unique'), the number of reads mapping to multiple locations ('multi') and the overall alignment rate. Percentages are based on the amount of reads sequenced ('all reads').\n\n")
}
  cat(ChIPhelper.Bowtie2(),sep="\n")
```


## Potential contaminants and rRNA content ##

A competitive mapping to the reference genome and known rRNAs (H. sapiens, M. musculus, C. elegans, D. melanogaster, D. rerio, X. tropicalis) as well as frequently occurring Mycoplasma species (M. arginini, M. fermentans, M. hominis, M. hyorhinis, M. orale, and Acholeplasma laidlawii), PhiX and ERCC spike-in sequences, E. coli and B. taurus (bovine) was performed with FastQScreen (http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/).

In the following plot only relevant contaminants are shown. A contaminant is considered relevant if it consumes `r SHINYREPS_FASTQSCREEN_PERC`% of reads or more in at least one sample. The "one genome" label refers to the amount of reads mapping to that genome only, no matter if they are mapping uniquely or multiple times. The "multiple genomes" label refers to the amount of read mapping to this genome but also to other genomes in the screening regardless of whether they are uniquely or multi-mapping. 

```{r fastqscreen_paragraph, echo=F, results='asis', fig.width=10, error=F, warning=F, message=F}
fastqscreen_plots <- ChIPhelper.fastqscreen(perc.to.plot=SHINYREPS_FASTQSCREEN_PERC)
```

```{r fastqscreen_paragraph2, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=(round((3*fastqscreen_plots$no.of.genomes)/4)+1)*fastqscreen_plots$no.of.samples/8+1}
  if(is.null(fastqscreen_plots$errortext)) {
    plot(fastqscreen_plots$p.category.wrap)
    } else {
      cat(fastqscreen_plots$errortext)
      }
```


## UCSC Genome Browser tracks ##

Browser coverage tracks are generated using the bamCoverage tool in [deepTools](https://deeptools.readthedocs.io/en/develop/) and they are normalized using the Counts Per Million mapped reads (CPM) method. The CPM formula is: CPM  = read counts per bin (typically 10 bp) / number of mapped reads (in millions). 

To view the tracks in the UCSC Genome Browser, navigate to the [UCSC Genome Browser Track Hub page](https://genome-euro.ucsc.edu/cgi-bin/hgHubConnect) and enter the following line into the `URL` field:

```{r Trackhub, echo=F, results='asis', error=F, warning=F, message=F}
cat(ChIPhelper.Trackhub(), sep="\n")
```


```{r samtoolscov_paragraph, echo=F,results='asis',error=F,warning=F,message=F, eval=file.exists(SHINYREPS_SAMTOOLSCOV_OUT)}
# mapping statistics for either paired end or single read design
cat("## Read distribution per chromosome ##\n\n")
cat("\nShow read distribution per chromosome generated by samtools coverage to inspect mitochondrial read content in ATAC-Seq.\n")

  cat(ChIPhelper.samtoolscov(),sep="\n")
```


```{r InsertSize, echo=F, error=F, fig.width=15, message=FALSE, warning=FALSE, results='asis', eval=isPaired}
# for paired end design only
cat("## Insert Size ##\n\n")

insert <- lapply(branches[1], function(dir) {
  if(length(branches)>1) {cat(paste("\n\n### Insert size metrics for bam files from", if(dir==""){"default"} else {dir}, "pipeline branch\n\n"))}

  if (!file.exists(file.path(SHINYREPS_INSERTSIZE, dir))) {
  	cat("Insert size data not available.\n")
  	} else {
  	  
  	cat(ChIPhelper.insertsize(subdir=dir), sep="\n")
  	
  	cat("\nThe insert size distribution is given as histogram:\n\n")  
  	 
  	p <- ChIPhelper.insertsize.plot(subdir=dir)
    ChIPhelper.subchunkify(p, 
                         fig_height = dim(p)[1] * 15/dim(p)[2],
                         fig_width = 15)
  	}
})
```

## PCR bottleneck coefficient ##

The PBC (PCR bottleneck coefficient) is an approximate measure of library complexity (without duplicate removal). Provisionally, 0-0.5 is severe bottlenecking, 0.5-0.8 is moderate bottlenecking, 0.8-0.9 is mild bottlenecking, while 0.9-1.0 is no bottlenecking.

Low values can indicate a technical problem, such as library complexity bottleneck. Nuclease-based assays (DNase, MNase) detecting features with base-pair resolution (transcription factor footprints, positioned nucleosomes) are expected to recover the same read multiple times, resulting in a lower PBC score for these assays. Note that the most complex library, random DNA, would approach 1.0, thus the very highest values can indicate technical problems with libraries. Some common numbers from ENCODE datasets are: 82% for TF ChIPs, 89% for Histone ChIPs, 77% for DNase, 98% for FAIRE, and 97% for control ENCODE datasets with no or mild bottlenecking.

```{r PBC_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.PBC(),sep="\n")
```


## IPstrength ##

To estimate the IP strength, we attempt to decompose the population of IP reads into two distinct components: those pulled down by the antibody, and background. The enrichment plot is useful to detect several forms of biases:

* The distance between the two curves depends on the enrichment of the IP versus the input; close curves suggest weak or diffuse enrichment.
* Under-sequencing produces curves close to the x-axis for a large percentage of bins since no reads were aligned on this bins. 
* Sequencing bias towards a small set of specific regions produces curves shifted towards the y-axis for a large percentage of bins.

For more details see Fig.2 of the CHANCE paper (https://doi.org/10.1186/gb-2012-13-10-r98).

```{r IPstrength_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
ipstrength <- lapply(branches, function(dir) {
  if(length(branches)>1) {cat(paste("\n### Fingerprint plots from", if(dir==""){"default"} else {dir}, "pipeline branch\n\n"))}
  cat(ChIPhelper.IPstrength(web=F, subdir=dir),sep="\n")
})
```


```{r PhantomPeak_paragraph, echo=F,results='asis',error=F,warning=F,message=F, eval=!isPaired}
# for single read design only

cat("## Cross-correlation analysis ##

**Cross-correlation:**
A measure of enrichment derived without dependence on prior determination of enriched regions. Forward and reverse strand read coverage signal tracks are computed (number of unique mapping read starts at each base in the genome on the + and - strand counted separately). The forward and reverse tracks are shifted towards and away from each other by incremental distances and for each shift, the Pearson correlation coefficient is computed. In this way, a cross-correlation profile is computed, representing the correlation between forward and reverse strand coverage at different shifts. The highest cross-correlation value is obtained at a strand shift equal to the predominant fragment length in the dataset as a result of clustering/enrichment of relative fixed-size fragments around the binding sites of the target factor or feature.

**Normalized Strand Cross-correlation coefficient (NSC):**
The NSC is the ratio of the maximal cross-correlation value (which occurs at strand shift equal to fragment length) divided by the background cross-correlation (minimum cross-correlation value over all possible strand shifts). Higher values indicate more enrichment, values less than 1.1 are relatively low NSC scores, and the minimum possible value is 1 (no enrichment). This score is sensitive to technical effects; for example, high-quality antibodies such as H3K4me3 and CTCF score well for all cell types and ENCODE production groups, and variation in enrichment in particular IPs is detected as stochastic variation. This score is also sensitive to biological effects; narrow marks score higher than broad marks (H3K4me3 vs H3K36me3, H3K27me3) for all cell types and ENCODE production groups, and features present in some individual cells, but not others, in a population are expected to have lower scores.

**Relative Strand Cross-correlation coefficient (RSC):**
The RSC is the ratio of the fragment-length cross-correlation value minus the background cross-correlation value, divided by the phantom-peak cross-correlation value minus the background cross-correlation value. The minimum possible value is 0 (no signal), highly enriched experiments have values greater than 1.\n\n")

phantompeaks <- lapply(branches, function(dir) {
  if(length(branches)>1) {cat(paste("\n### Cross-correlation plots from", if(dir==""){"default"} else {dir}, "pipeline branch\n\n"))}
  cat(ChIPhelper.PhantomPeak(web=F, subdir=dir), sep="\n")
})

```

## ChIP-seq peaks ##

The ENCODE consortium empirically identified 'blacklist regions' (BRs) for some species and genome versions. These blacklist regions are often found at specific types of repeats (eg. centromeres, telomeres and satellite repeats) and are troublesome for high throughput sequencing aligners. These regions also confuse peak callers and result in spurious signal in both ChIP and Input. Therefore, it is suggested to check the amount of peaks overlapping these regions and remove them for the downstream analysis. The table shows total number of peaks detected by MACS2, the number of peaks without overlapping the blacklist regions and the percentage of peaks overlapping these regions.  

Note that these blacklists are applicable to human (hg38 and hg19), mouse (mm10 and mm9), worm (ce10) and fly (dm3) genomes.

```{r BLACKLISTFILTER_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
blacklistfilt <- lapply(branches, function(peaksdir) {
  if(length(branches)>1) {cat(paste("\n\nBlacklist filter applied for peaks from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n\n"))}
  cat(ChIPhelper.BLACKLISTFILTER(subdir=peaksdir),sep="\n")
})
```

The number of peaks called at the default false-discovery rate (FDR) cutoff of 5% by MACS2 (https://github.com/taoliu/MACS/) in the different comparisons is shown below.

```{r PeakCall_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
peaksAllBranches <- lapply(branches, function(peaksdir) {

  peaks <- ChIPhelper.init("readPeaks", subdir=peaksdir)

  l <- lapply(peaks,function(x) {
      x <- table(x$chr)
      xx <- as.numeric(x)
      names(xx) <- names(x)
      xx
  })
  p <- Reduce(function(x,y) {
      z <- merge(x,y,all=T,by=0)
      colnames(z)[-1] <- paste0(colnames(z)[-1],".") # avoid warning of duplicated column names 'y.x', 'y.y'
      rownames(z) <- z[,1]
      z[,-1]
  },l)
  #colnames(p) <- 1:ncol(p)
  colnames(p) <- names(l)
  
  # print the legend
  #cat("",fill=T)
  #cat(kable(data.frame(column=1:length(l),sample=names(l))),sep="\n")
  #cat("",fill=T)
  
  if(length(branches)>1) {cat(paste("### Peaks from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n\n"))}
  # print the peaks table
  cat("",fill=T)
  cat(knitr::kable(p,row.names=T, format="markdown"),sep="\n")
  cat("",fill=T)

  return(peaks)
})
```


## Correlation of peaks ##

The overlap of peaks is shown in 'intersect' mode with replicates within each group highlighted for comparison. Overlaps are counted if they occur in the respective set irrespective if they occur in any other set. E.g. a combination 1 1 0 means a set of elements is in set A and B, and they can also in C or not in C. Under this mode, the combination sets can overlap. For mutually exclusive combination sets, i.e. overlaps are counted if they occur in the respective set but not in any other set (combination 1 1 0 means a set of elements is both in set A and B, while not in C) select Mode="distinct". By default, the number of overlapping peaks is displayed (it is also possible to instead display peak overlap based on bp). Those sets belonging exclusively to one experimental group are highlighted by color.

```{r peak_upset_paragraph_by_groups, echo=F,results='asis',error=F,warning=F,message=F}
# Display pre-calculated upset plots
displayPlots <- lapply(branches, function(dir) {
  if(length(branches)>1) {cat(paste("\n### Upset plots from", if(dir==""){"default"} else {dir}, "pipeline branch\n\n"))}
  cat(ChIPhelper.display.plots(web=F, plotdir=SHINYREPS_UPSETPLOT, subdir=dir, plots_column=1, addPlotLabel=F),sep="\n")
})

## alternatively, you can run the helper function for creating upset plots from here
## either with or without pre-calculated combination matrices (provided as parameter 'matrixlist').
## Mind that calculating the combination matrix for an UpSet Plot is quite time consuming (may take hours for huge datasets).
# upset <- lapply(branches, function(peaksdir) {
#   if(length(branches)>1) {cat(paste("### Peaks from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n\n"))}
#   if(file.exists(file.path(SHINYREPS_UPSETPLOT, peaksdir, "upsetPlot.RData"))) {
#     load(file.path(SHINYREPS_UPSETPLOT, peaksdir, "upsetPlot.RData")) # load object 'upsetPlotList'
#   }
#     ChIPhelper.UpSetPlot(peaksdir, Mode="intersect", peakOverlapMode=c("peaknumber"), targetsdf=targets, addBarAnnotation=F, matrixlist=NULL)
# })

```




```{r peak_venn_paragraph_combined, echo=F, results='asis', error=F, warning=F, message=F, fig.height= (nrow(targets)*2.1+5)/2.54}
if(length(branches)>1) {
  cat("\n\n### Overlap of peaks from different pipeline branches\n")  
    
  peakscovcomb <- lapply(branches, function(peaksdir) {
    x <- ChIPhelper.init("readPeaks", subdir=peaksdir, peaks_as="GRanges")
  })
  
  peakscovcombReorder <- list()
  for(b in names(peakscovcomb)) {
    for(p in names(peakscovcomb[[b]])) {
      peakscovcombReorder[[p]][[b]] <- data.frame()
      peakscovcombReorder[[p]][[b]] <- peakscovcomb[[b]][[p]]
    }
  }

  emptyPeakSets <- sapply(peakscovcombReorder, function(x) {any(sapply(x, function(y) {length(y)==0}))})
  peakscovcombReorder <- peakscovcombReorder[!emptyPeakSets] # exclude entries with empty peak sets

  anno_list <- list()
  for(p in names(peakscovcombReorder)) {
    anno_list[[p]] <- anno_barplot(comb_size(make_comb_mat(peakscovcombReorder[[p]], mode="distinct")), 
                                   gp=gpar(fill="steelblue"), height=unit(2, "cm"))
  }

  i <- which.max(sapply(anno_list, function(x) x@n))   # keep only contrasts which have representation in [filt, unfilt, filt*unfilt]
  anno_list <- anno_list[sapply(anno_list, function(x) x@n == anno_list[[i]]@n)]
  ha <- do.call(HeatmapAnnotation, anno_list) # gap=unit(2, "mm"), annotation_name_side="left", annotation_name_rot=0
  us1 <- make_comb_mat(peakscovcombReorder[[i]], mode=c("distinct"))
  UpSet(us1, top_annotation=ha, right_annotation=NULL, row_names_side="right", 
        column_title="Overlap of peaks from different pipeline branches",
        comb_col="steelblue", bg_col="#F0F0FF", height=unit(4, "cm") ) # height=3 # heatmap_height=unit(24, "cm")
}
```


## Functional annotation of peaks ##

The following plots show the feature distribution based on their chromosome region 

### Feature distribution of peaks for all samples ###

```{r feature_distribution_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
featureDist <- lapply(branches, function(peaksdir) {
  functionalPlot <- file.path( SHINYREPS_PEAK_ANNOTATION, peaksdir, "ChIPseq_Feature_Distribution_Barplot.png")
  if (file.exists(functionalPlot)) {
    if(length(branches)>1) {cat(paste("\n\n#### Feature distribution from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n"))}
    cat("![Peak_Annotation img](", functionalPlot, ")",fill=T)
  } else {cat("\nNo plot available\n")}
  return(functionalPlot)
})
```


### Distribution of TF-binding loci relative to TSS for all samples ### 

```{r feature_distribution_TSS_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
TFbindingDist <- lapply(branches, function(peaksdir) {
    functionalPlotTSS <- file.path( SHINYREPS_PEAK_ANNOTATION, peaksdir, "ChIPseq_Feature_Distribution_Related_to_TSS_Barplot.png")
    if (file.exists(functionalPlotTSS)) {
      if(length(branches)>1) {cat(paste("\n\n#### TF-binding loci distribution from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n"))}
      cat("![Peak_Annotation img](", functionalPlotTSS, ")",fill=T)
    } else {cat("\nNo plot available\n")}
    return(functionalPlotTSS)
})
```


### ChIP peaks coverage over Chromosomes ###  

```{r peak_annotation_coverage_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
peakscov <- lapply(branches, function(peaksdir) {
    if(length(branches)>1) {cat(paste("\n\n#### ChIP peaks coverage from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n"))}
    cat(ChIPhelper.peakAnnotationCoverage(web=F, subdir=peaksdir),sep="\n")
})
```


```{r peak_annotation_coverage_combined, echo=F, results='asis', error=F, fig.width= 10, warning=F, message=F, eval=F}
# covplot seems not to work anymore with lists or GRangesList in ChIPseeker 1.26.0 

if(length(branches)>1) {
cat("\n\n#### Overlap of peaks from different pipeline branches\n")  
  
peakscovcomb <- lapply(branches, function(peaksdir) {
  x <- ChIPhelper.init("readPeaks", subdir=peaksdir, peaks_as="GRanges")
})

peakscovcombReorder <- list()
for(b in names(peakscovcomb)) {
 for(p in names(peakscovcomb[[b]])) {
     peakscovcombReorder[[p]][[b]] <- data.frame()
     peakscovcombReorder[[p]][[b]] <- peakscovcomb[[b]][[p]]
     # sort by element size to print GRanges object with most peaks first and then objects with lesser amount of peaks
     peakscovcombReorder[[p]] <- peakscovcombReorder[[p]][order(sapply(peakscovcombReorder[[p]], length), decreasing=T)]
  }
}

plotcombined <- lapply(names(peakscovcombReorder), function(x) print(ChIPseeker::covplot(peakscovcombReorder[[x]], 
                                                                                         weightCol = "X.log10.pvalue.", 
                                                                                         title=x, ylab="-log10(pvalue)")))
}
```




### UpSet Plot ###

```{r peak_annotation_upset_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
UpSetplots <- lapply(branches, function(peaksdir) {
    if(length(branches)>1) {cat(paste("\n\n#### ChIP peaks from", if(peaksdir==""){"default"} else {peaksdir}, "pipeline branch\n"))}
    cat(ChIPhelper.peakAnnotationUpSet(web=F, subdir=peaksdir),sep="\n")
})
```



```{r Enrichment_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", eval=runEnrich}
cat("\n## Enrichment Analysis\n",
    "Enrichment Analysis is carried out using GREAT ",
    "(http://great.stanford.edu). It assigns each gene a regulatory domain ",
    "consisting of basal domain that extends 5 kb upstream and 1 kb downstream from ",
    "its transcription start site, and and an extension up to the basal regulatory ",
    "domain of the nearest upstream and downstream genes within 1 Mb. GREAT performs ",
    "both the binomial test over genomic regions and the hypergeometric test over ",
    "genes to provide an accurate picture of annotation enrichments for genomic ",
    "regions. The ontology terms reported below are ranked by the binomial FDR ",
    "q-value.\n\n",
    "Note that GREAT (> v4.0.0) supports recent assemblies of the ",
    "well-annotated human (UCSC hg19 and UCSC hg38) and mouse (UCSC mm9 and UCSC ",
    "mm10) genomes.\n\n")


resultsenrich <- lapply(branches, function(peaksdir) {
#    if(length(branches) > 1) {   # NOTE: regardless of the branches, we *cannot* skip level 3 section (###) 
      cat("### Enrichment results from", if(peaksdir=="") "default" else peaksdir, "pipeline branch {.tabset .tabset-pills}\n")
#    }

    for(i in 1:length(levels(factor(targets$group)))) {
    
        # subsection
        cat("#### ", levels(factor(targets$group))[i], " {.tabset .tabset-pills}\n", fill=T)
        cat("", fill=T)
    
        # GO Enrichment Plot
    
        cat("##### GO terms enriched (Biological Process, BP)", fill=T)
    
        GO_bar     <- file.path(SHINYREPS_GREAT, peaksdir, gsub("=.*", "", levels(factor(targets$group))[i]),"_GO_Barplot.png")
        GO_table   <- file.path(SHINYREPS_GREAT, peaksdir, gsub("=.*", "", levels(factor(targets$group))[i]), "_GO_Enrichment.csv")
        
        # if there are GO enrichment results (the barplot was created)
        if(file.exists(GO_bar)) {
    
            # GO enriched terms in a barplot
            cat("Listed are the top five enriched GO terms (BP) found using peaks with FDR < 0.01 detected in all replicates.", fill=T)
            cat("![GREAT_analysis img](", GO_bar, ")\n", fill=T)
        
            # and table of GO enriched terms
            cat("The table shows the top 10 enriched GO terms (BP):\n", fill=T)
            enrich  <- read.csv(file=GO_table, row.names=1)
            r       <- min(10, nrow(enrich))
            enrich  <- as.data.frame(enrich)
            enrich  <- cbind(enrich[1], enrich[2], enrich[8], enrich[9], enrich[6], enrich[5], enrich[7])
            colnames(enrich) <- c("GO ID","Term Name", "Binom Raw P-value", "Binom FDR Q-Val", "Binom Fold Enrichment", "Binom Observed Region Hits", "Binom Region Set Coverage" )
            #colnames(enrich)  <- gsub("GO.Biological.Process.", "",  colnames(enrich))
            #colnames(enrich)  <- gsub("\\.", " ", colnames(enrich))
            cat(kable(enrich[1:r,], row.names=T, align="r", format="markdown"),  sep="\n") # show only the top min or top 10 biological processes
    
            cat("\n", fill=T)
    
        } else {
            cat("no GO enrichment results available\n", fill=T)
        }
    
    
    
        # Pathway Enrichment Plot
        # up-regulated genes
        cat("##### Pathways enriched (Molecular Signatures Database, MSigDB)", fill=T)
    
        Pathway_bar     <- file.path(SHINYREPS_GREAT, peaksdir, "/", gsub("=.*", "", levels(factor(targets$group))[i]), "_MSigDB_Pathway_Barplot.png")
        Pathway_table   <- file.path(SHINYREPS_GREAT, peaksdir, "/", gsub("=.*", "", levels(factor(targets$group))[i]), "_MSigDB_Pathway_Enrichment.csv")
       
        # if there are Pathway enrichment results (the barplot was created)
        if(file.exists(Pathway_bar)) {
    
            # Pathways enriched in a barplot
            cat("Listed are the top five enriched MSigDB Pathways found using peaks with FDR < 0.01 detected in all replicates.", fill=T)
            cat("![GREAT_analysis img](", Pathway_bar, ")\n", fill=T)
        
            # and table of Pathways enriched
            cat("The table shows the top 10 enriched MSigDB Pathways:\n", fill=T)
            enrich  <- read.csv(file=Pathway_table, row.names=1)
            r       <- min(10, nrow(enrich))
            enrich  <- as.data.frame(enrich)
            enrich$MSigDB.Pathway.ID <- ifelse(nchar(as.character(enrich$MSigDB.Pathway.ID)) > 10, paste0(substr(enrich$MSigDB.Pathway.ID, 0, 15), "..."), enrich$MSigDB.Pathway.ID) # shorten long lines
            enrich  <- cbind(enrich[1], enrich[2], enrich[8], enrich[9], enrich[6], enrich[5], enrich[7])
            colnames(enrich) <- c("Pathway ID","Term Name", "Binom Raw P-value", "Binom FDR Q-Val", "Binom Fold Enrichment", "Binom Observed Region Hits", "Binom Region Set Coverage" )
            #colnames(enrich)  <- gsub("\\.", " ", colnames(enrich))
            cat(kable(enrich[1:r,], row.names=T, align="r", format="markdown"), sep="\n") # show only the top min or top 10 pathways
            cat("\n", fill=T)
    
        } else {
            cat("no Pathway enrichment results available\n", fill=T)
        }

        cat("#### {-}\n", fill = T)
    }
    cat("### {-}\n", fill = T)
})
```

```{r diffbind, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", fig.height=4, fig.width=10, eval=runDiffbind}
cat("\n## Differential binding analysis\n\n")
cat("The DiffBind package ",
"(https://bioconductor.org/packages/release/bioc/html/DiffBind.html) is used for ",
"identifying sites that are statistically significantly differentially bound ",
"between sample groups based on evidence of binding affinity (measured by ",
"differences in read densities).\n ",
"After the occupancy data obtained from the peak caller (peak sets) is read in, ",
"a merging function finds all overlapping peaks and derives a single set of ",
"unique genomic intervals covering all the supplied peaks (a consensus peakset ",
"for the experiment). Then the supplied sequence read files (bam files) are used ",
"to count how many reads overlap each interval for each unique sample (refered ",
"to as 'fullLibSize', see below). The proportion of reads for each sample that ",
"overlap a peak in the consensus peakset is reported as 'FRiP' (Fraction of ",
"Reads in Peaks, see below), and can be used to indicate which samples show more ",
"enrichment overall. The total number of reads in consensus peaks is given as ",
"'PeakReads' in the table below. Optionally, e.g. for narrow peaks obtained from ",
"transcription factor binding, the peaks in the consensus peakset may be ",
"re-centered and trimmed based on calculating their summits (point of greatest ",
"read overlap) in order to provide more standardized peak intervals.\n ",
"The applied normalization procedures, the calculated normalization factors ",
"('normFactors') and the normalized library size ('normLibSize') are given in ",
"the tables below. The final result of counting is a binding affinity matrix ",
"containing a (normalized) read count for each sample at every potential binding ",
"site (affinity scores). This matrix is used for further QC plotting and ",
"differential binding analysis for the contrasts defined in ",
"'contrasts_diffbind.txt'. For this, DiffBind uses statistical routines ",
"developed in an RNA-Seq context (primarily the Bioconductor packages edgeR and ",
"DESeq2). It will assign a p-value and FDR to each candidate binding site ",
"indicating confidence that they are differentially bound given the respective ",
"contrast.\n\n")

diffplots <- lapply(branches, function(dir) {
    if(length(branches)>1) {
      cat("### Diffbind results from", if(dir=="") "default" else dir, "pipeline branch {.tabset .tabset-pills}\n")
    }
    ChIPhelper.diffbind(subdir=dir)
    cat("### {-}\n", fill = T)
})
```


## Used tools and versions for this analysis ##
```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat("\n", fill=TRUE)
cat(Toolhelper.ToolVersions(), sep="\n")
cat("\n", fill=TRUE)
```

### R session info for generation of this report

```{r R_sessionInfo, echo=F, error=F, warning=F, message=F}
cat("\n", fill=TRUE)
sessionInfo()
cat("\n", fill=TRUE)
```


</div>
