# ChIP-Seq pipeline
Here we provide the tools to perform paired end or single read ChIP-Seq analysis including raw data quality control, optional adapter trimming, read mapping, peak calling, differential binding analysis and functional annotation. With adjusted parameter settings, this pipeline can also be used for related assays like CUT&Tag, CUT&Run and ATAC-Seq. As input files you may use either zipped fastq-files (*.fastq.gz*) or mapped read data (*.bam* files). In case of paired end reads, corresponding fastq files should be named using *.R1.fastq.gz* and *.R2.fastq.gz* suffixes. The Pipeline performes 2 branches for bam file processing in parallel:
- filtered branch: removes multimapping reads (MapQ filtered) as habitually done in most ChIP-seq studies and (optionally) removes duplicates from BAM file assuming they are unwanted PCR artifacts (duplicate removal is usually recommended for paired-end data but not for single-end data).
- unfiltered branch: alternative workflow using the unfiltered BAM files. This may be preferable when studying some types of repetitive regions (keep ESSENTIAL_DUP="auto" for MACS2 peak calling).


## Pipeline Workflow
All analysis steps are illustrated in the pipeline [flowchart](https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=NGSpipe2go_ChIPseq_pipeline.html#R7R1Xd5vI%2BtfobPIgHYrqo5uc7M16HdvZlBcfBCNp1ggwIJf8%2BjsVUQY0kmiyNzknEcMAU775eunoZ6uXS9%2Fwln%2B5FrA7mmK9dPTzjqap%2BlBH%2F%2BGWV9ai9Pu0ZeFDi7VtGm7hb8A7stY1tECQ6Bi6rh1CL9louo4DzDDRZvi%2B%2B5zsNnft5Fc9YwEyDbemYWdbv0MrXLJWdTjZ3PgE4GLJPj3WRvTGzDAfFr67dtj3HNcB9M7K4K9hcwyWhuU%2Bx5r0i45%2B5rtuSH%2BtXs6AjdeVrxh9bppzNxqyD5xQ5oHzf1%2BD82%2F%2FPD78Xp6Nrucq%2BBwq3Whw4StfC2ChpWGXrh8u3YXrGPbFpvWUzBfg1yro6t%2F1yuP9F4aHWjZPfXFd1HCu4m4gDF%2FZzhvr0EVNy3Bls7vgBYY%2F8At7A3b1M3bn%2FIV9i1y88gsn9F9jD%2BHLn%2FF7m8fIFX8uCA0%2FZCMZ4mt37ZvgGvhwBULg33qGCZ1F9KgV64qeXIAwp2sQ%2Bu5DBD9of0%2BzW8R2jX6SNj2ZV19vv6%2F%2FXJ%2F%2F%2FnX3%2BW76M3j61dUZ4NEPFnTssw3Euxb7BAOBS%2BCigfqvqIMPbCOET0mAN9i5WUT9NgCEfjAYEsPTJ%2BXP%2BfJqHNpPy5vZ1afb6f3vT11%2BoJ8Me80%2B1dGGNprBqQWf8BdtuHDIjeHjGoP%2BqU%2BOVXSJfi3Y%2F%2BSxmZ9uQQMj7%2BKtKfiNQZVtzIB9Gh3RM9d2fdJJn5I%2Fwk2buw6HDhXvuWUESwLsagL08dUc2nb8pQr%2BG72U3yE4QT9d%2BIYFESCkmk13BU0GQQvbCAIOTRxdKOSzIdo8Fy9cd6IUQdYT8EPwUggJ7O54qPe0AX2K4e2uqqpR23MMEY7Yti5jOLCv9HuDw8FICNXqu4ajiyH%2BWwUcwRUiZCeBR2moWg4gaWM1AUVjJQNAfWWSBaARx12HgM%2B%2F86vgy%2Fzl5fVm%2FQtqI%2FjjTvmnOziYqNVNwtQEAcNncBsNK5HWZEmNcFXVpiiNcDTDo9tjJbnHerNbLO6nN7XHRcPOIwIpBJ9uwAg4ASGcNOAb3YDs8QnqoPa9FwHdONM6J3FaQV%2BXISFlfzY9m7Pl5%2BvuLXhET1xDD9gQYXs0tNPCMQiJXc4EBEtJSeyBMzhsHboECmFozNDTCOj51JGQCebQgYQbQuCD%2F9kyx3I2xMDUywfzxBuWYYgl1BMM5tp0AUPEKfR82%2Bs56JhpU7iazV105NDPq8tbPAdt4aKLLt4z253hQ2EESJBAP%2FgMA%2FQb73mAt3xqLpEIDB57%2FG4PcSDu02t2mPkd%2BdIYkiBRznIFnuHItKFB7NZcMxS3%2BoDYrmEFWAuBdx3rHFxrbWMIUgzHwgcoCBCpgIZNyYZBhNaScEQJx%2FsFmOuQDDd2vNGBwa9FYjU5%2BB8C8BH9Rxo9A%2FqAzIvM7oNHblkgwKKA%2FKzy%2BfxSmGElwQxnham%2BKhCmRiWwwmL9gICEphYASQge%2FokkCA%2FJEnjqlA3ivMeYN0QqsL5wsQppuPQKqpkV0wQLxtukGRH2vWsXErhlH%2BumJOBR8gXufB5gTJ7ahGjQB0i4jcoo%2FPfPBC%2B7G%2F9ap4iiarIMbE3867cA%2BH%2FP%2FsVyNELDWD0gUlJkkePMfcG4kSgNMXacub4F%2FC5qxnwMXiIO%2BAR3KgjKB9GduLL5hPA9Oujjv1EPz7Cs6N1a%2B9gZun9BL3wRaGqSN4vYFgE54oxidsnz1zNNqcAqPijFRLsdRBrezS0l0yta9lgvLX%2Ftw5lrvWYaM4QrtHjL5%2BtC8rbpmH2Ut0DeMF8T5IRfqQTGyiOr5iDWIPZCKPUR3JgddOE04t85fCpkDvSVSohwTTETUNWkrq6%2F3ZW%2FPY63DvN2qPbplbxtbG5N7xxGpd6%2Bk2AP80m489hmVT3wG%2BDZ0DTCEvfEWa9mGG%2FTifjpD4jmIvnRnWd3DYyHM8O2t0gqu00vtkseej%2FG24lPlD1B3CjC8oQGEkoVl0aKOUUtxsIw1jLGNWLGGu2VfcIMIiFmBU%2B5ecQGc%2FwZF%2FWa28QMgG0NnQRTWIbUk7QAqFrWBKApAi5eHUx2Zr3QZYz72oHl1sb%2FsdzyLDcXFI%2BG5faE3N8KDR86lMFTvA2XTdu75LTge8PYPQT0YZedoBPC4jlUb7JNV3u9UWNk1C3Salov3bbM9NprqhQV4JtjsWBw2tmoUnfQyYnampEuInVX78nwg1xVaV63rMTBpKyCzcpZp8weVrNjKSrn%2Be4TtMA2xd82WncaaRKv0IO9JTCQiJpcC8En6RNdbFuGc2gmP6x8cL2w93Fnahtbx610Uiumk7tQxOclDAF27sHPPvvEqal8KtnXR3JUcrK7m8X%2BRFJkc6tVYcg%2F2FqNYT%2B5ixVoDIWueqqa2Yg3x77U5ZiX8MXZn0fSpe3qw7KZJPLoie8br7EOHgbVIB92R2rSOKEyffdUsr8%2B0VKQTUcgCefbOLiIP5oiOv%2F1jJAcGebpcY2Qe4gXwnQRQLl2JMcaz2RNiZUMzwPxDA8dZvUpjRmJrG64K7Y2oRuPJqeb2wkXhaJcwiXCtilixrSUX8jVuT7MIWHEI429WCldBhwoowSwTLLUTR0K8PJwd9OXDHX7%2FT08v1N%2B%2FPpzenPhr1Tr7vL1c1edFOJQYM%2Fc58bRZ3Xoqi%2BJrgalOxXvha76KVtqn0nwe6KfDIhMb4bm5Y%2Fz5Q348mjOf77e%2FX32u9uokkBJgMloLypbq1%2BZLETxeJD6HcuEm6zqje5yLUEO1e6ycFlL53IO2mT98E0uNbYlSQNG7VcBSp9uHq3SNL0YJumFxrzO89jbdH9dKZe%2B%2BLcX10t1th7Ovz2ew%2FNT9%2F4EdhvFPGVIcbvTlwrZF%2BHBLx0aZdGQcMP7pWIhFu3RchhI7HgzDEe7gKBY7vgPCCoCglGrgEA9PGblPyjYAwoGNUGBtH5r5j6HEGjSCi7eEvN7uwGGha0uhudhRecWM1WeCztt%2BhAQ5whsp%2FFohAX2qP6Y93wNirSNTY8t1Q7atH5sc6vUppXsQcFUq5FqdsjAuyH1WaEaWxCQtd3ASeIAmJlT2cPa%2Fvc6pH5jOLsDM0imzhW6%2BWgWHanmzOsUtvExi6YBaagGDXjYjFvGEHq4%2F088OplHK0%2BNFbQx%2FH0C9hPAb%2B3UYRYd9cdJMWyoZUGfH48DQV8Szkctg3MlA%2Bo%2BCNY2FobbCOwfEJChuQHrYyfyt2PgXq%2BLxDmcz4G%2FcYWgQ7FhENY%2BlGv6bcNxeC6Hlp3uCk5yfyjIJSE6yWOtqpM8bv1Jtt1FS48xGlmXUynTtW3ELZDwO0qxgGEuO9zv5x0A81CXBOYyyNL%2Fni7vvbO7b%2BHVCgyvV%2Bur8P7q%2BFWH25TX%2BwuDPJXIVmFw0piNonDcrcFPWY7aBx4Cm5aiqKTXatVupzgnHRZPoyWZBkvovN6z4Pt72o5kx3vck8mxtLF3s7Kyw8%2Fcz%2Fqg5jhotnrKPdIH3UJiu%2B0BP%2BjdiOcu7pi7CI151uJNQlgpOg34zPjuqpOMXI9zuNKBTM1MCFFtOH8lxCzEVNxcByGdD3fa3WQRCBF8BUuEInATxH5VDgCYpkhP7M2LrCNVkjdQlRJSpk1FRq%2F%2B4e6hFVq7q%2BEXKtMt94%2FcsVNj2sO2OXZ%2BdlBPdogldd94%2FXGeH%2Fwg3jr8LMe%2BQ2OFodSZBfRcKzNjxRCxSNKiCUNEH2WrUYt7KJ0Knoe8Urtfl4to7CCnkhIC8qeTTkqI7uhDfaJbnViqwYo8SjW9L8xOWaNaXOyZf7iJ%2F7gQbfvz53Kb37EicL0%2FbCUC95aGgxhFokiVxeBfzzCfir4NzYCiZYq26b8Ya0PHokZMc4nYTxNBBgzCqHcKiQd5SLxWu2RsIXZA44P%2F0DiCVe4H0RAO704vgDa%2Bur369udfq%2B%2BnP0dQvWw2ZWwZKHy0r2doCvPtiW%2BHsvh21Ah61ZLocsikpVz0muo%2FSAZKVYZeL28uTu6kEeuFg1DqcgXIyA0EeK8BDDLvLNRINIU%2ByUR3QJzD%2BhFnAg3ujkUrQJxDHgLappAq7RhCqsrzpi%2FJl5l7Jm7Flzy3etP8qZ5CoMliI5UhRCTQn2FNIKnfIokWTR%2BQ1EWK4%2ForhAYCYrScwcUzxDxmiDjMhzrjQdEcTDaHHTBenE4ejxtbGmUN9GZ5ve%2BO%2FvB8%2Bnj36%2Fv1%2FY3zP0%2F%2F%2FHTfrB312IK%2FuKY7jqmEq9pYeQChOpxncn1L6vC9Wfyy9r3QWa5lwV%2BDYXFug0z%2Fmlh8rDG4j3mBSVM1hOkNHwYk7T01zD3BAGc%2B%2BM3cyerVfySnEWVR2k7Yxg2z8kmW3QJzY02WGpsTBa2p0jzkTzm0Mu3yPRDkA2pet601SSz%2F020XhLYdq267z6LvW6fbnpnUheLD2sEOzD0sNfiGYy5zFM5pFH19doMeOXXD0AYOMLG775kL5nNoQqKjqRdBz3ZIQUNB6t0rpsd6CxHw0Wum3xwC5ulxjhYBj9vpHQI9tEvAWbB0%2FTIo17Nd%2FFaazZ4%2FXC%2Bi3YxaHt8OlE7t%2BLYChNkfNBukKFRI68dX%2FbFJ9Y50FPW4bFx2mH7n8Dxd71y%2FI53cp536nb66W3KfA%2FU7cl4ER1iUtBzUU44TgQAkhcvcPxAAD83fsDLMQD57Q6zgBPFyqpU7IUPdQRB8G9kT9PGkUcak0JTbmlivrHMdt422MdRrBhffEwZcFofKtOBBIqi6vUkUSoD2dPQJN3XUEpk6G62tS%2FfqhwdW3a83k3BiXg6P33uuTs5rLM15NcVxFw57L%2FRVeemVG0ASPZjY7%2BMSOG6inp2cL1yjIaynf3%2B%2F%2B3xB6DM6FS%2FEZWWRmEc6ApTfVXC%2BcRwGigj%2Bmq5A3cGFZALQxUU%2F4vkt0OkwH7DnDi2TVhwb2XTJFBY0ZLsmGT2JdcRz2FYtZq9CJi2rS5JknkYsDWqSnEyy5ERVK8M1AwGuSS1ctVVIImz3fquQCHNQNBrqlJRwI%2BK%2Bk4Sr1kjpJwKveKE2a1BXikK54zfJz%2FPW%2BsLGEa2P5VYgoFQs1Ly3ksJ0SZLZ6NB%2FHxCQWzhzVyfKcJDOUcFTBjw4MPQ%2Fbq7n0CFJtqLMZLQGi0HATVSihToxbeIW0EIDL%2BjFNkWmxmdVZUh3fP87GVhnp7qmpZYprYj5mQzVJPcjTOUgql1aWSqHLK%2FTmDR9BFUKVFU61rdVRDYadyP6QF6Euzj1k%2FFsGVi2bKM%2BkJTZ6mzyqH74DT2PpJxsWKZEsi6LZnadJyw7ElmYKis9A1LiChiv08LcQLSMudq7UXtkjXuL3%2B0ep7bHOI9aAawlSVbW1iEuI7p7rW3JJJuTBhGZnGED50lPZKxqESLbJK2R1pAdNfSmzRcaL%2FO2zXwxqIzhOtx%2B0b5KUbV6k0xUSR6MJ7xsTzEGkoLwhqU3lLXpswXBNJ5qjwl1R%2Bf9wXKfnc4mTyAedW3Gfp5NEU9F3uY%2Fie%2FA8dr8R%2F1mbf6flD%2Fny6txaD8tb2ZXn26n978%2F8aBIAWXMABfGvkLqktYp9UVE5AtYYJZyA7z0ddsSPsT2GRPc5CYlPfFZuZm42z5r4tDASLR%2BmqZHK2hZBGeKoCKJR8ugL9w%2Fm5fSyJKXvog%2F0naHCzkgyJZUn%2BGz3Yklz%2B5o%2BstJh2bOt4mAQBJtuh6WHgz7Y2a39juVZaxuqiirKMC7L1jdflWre4hZephmPmNaWL6am0ayriK5JJGGMJYRdrM5jTOZe3KWuyjoygCu0SQBXGOBY4sQuErgDIXAVacdshC6368ZUrgsR2igYvrxKbW0FJzG92aKKlyMd2tpae3AGjQBlUBhJqP92Be9n48bD6IwTZaagtQIMHVTOrRWesm9A81YGjil2Z9hRcCZXz0pUq9%2FDtFqmnGf8Y3mnWznOsAQpTBYi2oHvFnpc5TUbqqyGKYy8TOrnT%2BnEddEMwWop6OFVowkTRN5ahDHjLe8ZUkuVu1nt0wTblkJFhWhP%2F2bTyi%2Bp%2B5ZuFiyque6zP%2FbNM8t8Ja%2FjusonDlcrH3u9dxGC1my5BGSWhzB2ETNVVdHoouHb%2BB%2BPXq58RFPG%2FqEnSK2ZiqYVIMlkOJLXts6ej4wVkiICHIXUdCjBYuVYoqSqyfyRqh6RbkWMmkgItW8cpc2dbeNy5pqsMCcYrFn13%2BwXVYS3TEWhOUsWn9R2yENPqB0Ktjhux%2FKGyBPWIf4CjYCtpaBvfYTih9cjIEW8XLj7LgSbP8IqcQDXoCJWAc%2Fsc84asZObKDUvLD3MPUCFt3dxOZExbOjwx89SKvOJcJ2PhBb7AMAGCY3CxMg5gc6i%2BBjb8vwUtJkIaNWbDptgXtrylVowNnYbc5CakVWUuEqagIxsyotexEL%2BY617FNhTbni3O5HmWSlTslEOoVu6bl1yKOHJnkfsVRKuXnC0knhh4MUFFaTJ4zVoibstLzLjul6r2QzAlIDnlTdRuSjEwuCYY1cC6p1KvfdWcEAm1gMB7jrgDDCsan1aL%2BeOH9HPsQdnStPOq%2FYSGm2ypsYGWaJUqPIUOnthA7RRZQaMY0i1eZQpCh4Q7j83IunJfnHeIXQtw8NWtPgUOhw2hZwaFk6uhLBIcU%2F6Q1DQxGjVT8wFA5bSp5Z%2B%2FbrKc5ohCe4jdZuIIZe8RoC593JjmsvT5z7o0FvkJQ%2FJnpPFCSZpc%2F6qLL8j29QNKnKY7%2BoXMuxpn8cMnasVekfDycCTQeN7Ivdd83%2FeBBIJqoBNhEyYsH5HBetlZY98QM4PRXOI9SJF7zdZD%2BoQdjc2AH4BHShZJm%2FOfVKlrtVaimq11hl2Rb5qoy7Q%2BTewmqzZVtanARbuFqyaXpKJ4WHsUDqwXvcOhaoIg5IdtcLw5tbxgFpw90SYOssrVvV2lmEemfucwiBtnbgIxaAJImkD1buEzH3IYQOuzhkmxJJH9EocS14qTI9tVLW9OyFBDYfHGsmsA2RTI0fqTbpd9XD0ek7Ipmy9Yt5Tve20MyWKfGPjWbKbntLaabeL658VDLNzMDkv%2FOr4Mv85eX1Zv0LaiP44075p1UlFvfNNbEFBFMlvtLwGD9SzEW8BszUFn%2FgiFW5WZ2vPXmDdsQvWWvPJu771lEwS7Qqur9CwxZnqcjf1nrZI%2FQyaHIk2hCrpKvNltjyby%2Bul%2BpsPZx%2FezyH56fu%2FQksId2%2F4MS3nIpuQWH746uJLEltZzFBNVkcsDXFBE%2BN1dczaVwqinoi6Ws3yblqxI3oo4%2Fi6q35EHT0Tj%2FqUGsfpiuh%2BmqVqE6rS2JIYL%2F9cR3HBceK7IaThBWxNcjuL6wx2wHdBevVyvAxrGnK17NOlH2wNgxHVHy74Dj1OCumZoxF42aRnHBtB%2FmJiGk2lWwqJJ%2BMdLco%2FURekxRSja06gfPTKOFKjM%2Bekj9C%2FBQP2B%2Bm8VXSaybNvA%2Fx305O8PLCNyxcvD3VHJcMFrYRBGkpoaTI5ElGfyFKlSlKV1FZrszBttQ7O%2BTNIk%2Bt7XSLvYmqMp6xfEtyUAsj03H953jQqp0J6Im9DJD4sNCHK5agGQfkPEHwTGDR2LxICL4Sr96YLPDrYBAmEyMUvQFj3U1Nu29ntxgr80JWCgLhZ5pGISp%2BRzzXWe07gsfZGEw8KVJ0ac%2B5QAd9iQRM0fdmp7LzKw3TBEFAV8aGM9%2Fw6S7i%2FM0vdEtn%2BJ%2Frsxv80w1DGzjAJFUzXTCfQxMmk6gVfQxzDY4PzeWKPMNBhK1sfCipquFSbzd9NwjIuHzKetBCV%2BwbmU3xloYTkjIauAqoLDDgUqFEmcMeUnj%2BJxrz9g9w8EctaCD0tNrMjSbEZnBtG57s7iRKdiVWi5cujWCKjcBCAOHD2Zo9dDftxJ1abNeE0geHFwuhH5RwlJF6LymOBk1yLBYsLjABFen3ScI0asziqy207fA8GYyDOSBLRkJ9FadeZRCpvpoiUlnn0CgJZzKn0uhwIiVUqTeqUW%2BxKU9sf5CVyRoz5YmH3bIIjDYXTpHdd3G%2F0kMt9rPlTVKKxy3Rien%2BmlaP%2F8vZOjQswwulhXHSmzF3nDWtTRA32WiFkng%2B2NQgiBORjr1YSR65PPtLgspVIL1PBNJXjcJ74elscfENAtZtrb4RDU5JFRR6B%2FkGM5U4FJF6QcsCeBnaBWEExeE0veVmxLICKISrp8lyctXwbdtIZGtTMx%2FNAJvJGkXUbahLcM%2FjRnrhi0Cjl9uvKJOUINu1uPVN58DmSxflJtrQHVw6vZAaySR8hsm3Uds2%2F2iHKa7WAb4F4%2BVrCj4r0o2UkNt6M6iyphybJskXBpnWCrKyfCsD3%2FrjAz6kntol%2F2kf%2F2hi8is0Fh%2B%2BlDd3CwSY9aDzXNtGE5MK1rN78OIJQHv%2FWbG0b%2BjVXfRqErruVAaunZ0ynhfyN1qh8FRlkdwMVyhgMfIFoaSGTxuPsxo%2BYRYxnjSlZElIuLajY%2Bcd947%2FLoV57MsyjyMxqNTryd2f6MXan3Ljv6W1PyvjAdwvfPBqQ0xxEIeF2D%2F8%2Bx673yYq7RZrhVh3tEIIJ8wN8uHIllGDemjD%2FiWmJNQT5cNTvQ4beXqhnIDhnPBigRbpMOSZ0iL1h3qdaiTU5LtuGId0tHDLv1wLA%2FHF%2FwE%3D). Specify the desired analysis details for your data in the *essential.vars.groovy* file and if necessary in the individual module header files (see below). Run the pipeline *chipseq.pipeline.groovy* as described [here](https://gitlab.rlp.net/imbforge/NGSpipe2go/-/blob/master/README.md). A markdown file *ChIPreport.Rmd* will be generated in the output reports folder after running the pipeline. Subsequently, the *ChIPreport.Rmd* file can be converted to a final html report using the *knitr* R-package.


### The pipelines includes
- raw data quality control with FastQC, BamQC and MultiQC.
- adapter trimming with Cutadapt (optional)
- mapping reads or read pairs to the reference genome using bowtie2.
- filter out multimapping reads from bowtie2 output with samtools (performed in parallel pipeline branch).
- identify and remove duplicate reads with bamutil dedup (performed in parallel pipeline branch). 
- generation of bigWig tracks for visualisation of alignment with deeptools bamCoverage. For single end design, reads are extended to the average fragment size.
- characterization of insert size using Picard CollectInsertSizeMetrics (for paired end libraries only).
- characterize library complexity by PCR Bottleneck Coefficient using the GenomicAlignments R-package (for single read libraries only). 
- characterize phantom peaks by cross correlation analysis using the spp R-package (for single read libraries only).
- peak calling of IP samples vs. corresponding input controls (if available) using MACS2.
- apply a pre-defined BED file to filter out artifact regions or create and apply an individual grey list from provided control samples (optional).
- peak annotation using the ChIPseeker R-package (optional).
- differential binding analysis using the diffbind R-package (optional). For this, contrasts of interest must be given in *NGSpipe2go/pipelines/ChIPseq/contrasts_diffbind.txt* (see below).
- genomic regions enrichment analysis


### Filtering of artifact regions
Problematic artifact regions with unspecific signal can be excluded from analysis. Pre-defined BED files with problematic regions for your genome of interest (e.g. identified by the ENCODE project) can be specified as *.bed* file in *ESSENTIAL_EXCLUDEDREGIONS* (see below). The MACS2 generated peak files will be filtered for these regions and stored with an *_excludedRegions_filtered* suffix within their file name. Alternatively, or when no pre-defined BED file is available for your genome, you can generate your own "greylist" specific for your experiment based on your provided control samples as described in the [GreyListChIP](https://bioconductor.org/packages/release/bioc/vignettes/GreyListChIP/inst/doc/GreyList-demo.pdf) vignette. For this, set *RUN_MAKE_GREYLIST* in *essential.vars.groovy* to *true*. This greylist is then used to filter the peak files. IMPORTANT NOTE: if you create your own greylist, a potentially provided BED file in *ESSENTIAL_EXCLUDEDREGIONS* is ignored. By default, *RUN_MAKE_GREYLIST* is selected and no BED file is specified in *ESSENTIAL_EXCLUDEDREGIONS*.


### Differential binding analysis with DiffBind
Since version 3, DiffBind includes the data from ALL samples in a single model. If you have an experiment which involves different antibody pull-downs you may want to have the samples of the respective contrasts normalized and analyzed separately. In this case specify contrasts belonging to the same antibody pull-down by assigning them the same *sub_experiment* name in *contrasts_diffbind.txt*. The sub-experiments are processed isolated from each other.
DiffBind offers several normalization methods (given in *ESSENTIAL_DIFFBIND_NORM* below) and read sets to apply them on, e.g. full library size, reads in peaks or reads in background bins (see *ESSENTIAL_DIFFBIND_LIBRARY* and *background* parameter in *diffbind3.header*).
Note that DiffBind works with implicit "default" parameters, which depend on the context of other parameter settings (see [DiffBind documentation](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf) for explanation). But you can explicitly specify them as well in the *essential.vars.groovy* and *diffbind3.header* files, respectively. Your specified settings as well as potentially derived settings by DiffBind will be listed as an overview within the final html report.


### Pipeline-specific parameter settings
- targets.txt: tab-separated txt-file giving information about the analysed samples. The following columns are required
  - IP: full IP sample name without any filetype suffixes like ".R1.fastq.gz" or ".bam" (sample name itself must not contain dots).
  - IPname: (shortened) IP sample name to be used in plots and tables.
  - INPUT: full input control sample name without any filetype suffixes like ".R1.fastq.gz" or ".bam" (sample name itself must not contain dots).
  - INPUTname: (shortened) input sample name to be used in plots and tables. 
  - group: variable for sample grouping (Condition).
  - Replicate: number of replicate (needed if differential binding analysis is selected).
  - PeakCaller: name of peak caller (in this pipeline it is macs; needed if differential binding analysis is selected).
  - optional further columns which may be used in multifactor model design used in the model matrix design formula of the contrasts_diffbind.txt file: 'Tissue', 'Factor', 'Treatment'. 

- essential.vars.groovy: essential parameter describing the experiment including
  - ESSENTIAL_PROJECT: your project folder name.
  - ESSENTIAL_THREADS: number of threads for parallel tasks.
  - ESSENTIAL_SAMPLE_PREFIX: sample name prefix to be trimmed in the results.
  - ESSENTIAL_BOWTIE_REF: full path to bowtie2 indexed reference genome.
  - ESSENTIAL_BOWTIE_GENOME: full path to the reference genome FASTA file.
  - ESSENTIAL_PAIRED: either paired end ("yes") or single read ("no") design.
  - ESSENTIAL_READLEN: read length of library.
  - ESSENTIAL_FRAGLEN: mean length of library inserts.
  - ESSENTIAL_BSGENOME: Bioconductor genome sequence annotation package.
  - ESSENTIAL_TXDB: Bioconductor transcript-related annotation package.
  - ESSENTIAL_ANNODB: Bioconductor genome annotation package.
  - ESSENTIAL_DB: UCSC assembly version for GREAT analysis (only for UCSC hg19, hg38, mm9 and mm10)
  - ESSENTIAL_EXCLUDEDREGIONS: path to bed file with problematic regions to be excluded from analysis (optional). IMPORTANT NOTE: ESSENTIAL_EXCLUDEDREGIONS is ignored if a grey list is generated and applied (i.e. *RUN_MAKE_GREYLIST* is set to true). 
  - ESSENTIAL_ADAPTER_SEQUENCE: adapter sequence to trim with Cutadapt (optional)
  - ESSENTIAL_BASEQUALCUTOFF: base quality threshold to trim low-quality ends from reads with Cutadapt. If *ESSENTIAL_NEXTSEQTRIM* is true, qualities of terminal G bases are ignored. To switch off base quality trimming in Cutadapt entirely, set *ESSENTIAL_BASEQUALCUTOFF=0* and *ESSENTIAL_NEXTSEQTRIM=false*.
  - ESSENTIAL_NEXTSEQTRIM: most Illumina instruments use a two-color chemistry like the NextSeq (exceptions: MiSeq, HiSeq). This option accounts for terminal high quality G bases incorporated by faulty dark cycles during base quality trimming with Cutadapt.
  - ESSENTIAL_MACS2_BROAD: use broad setting for broad peak calling in MACS2 (default false).
  - ESSENTIAL_DEDUPLICATION: remove duplicated reads in filtered branch (default false is strongly recommended for single end data). 
  - ESSENTIAL_DUP: how MACS2 deals with duplicated reads or fragments.
  - ESSENTIAL_MACS2_GSIZE: mapable genome size for MACS2
  - ESSENTIAL_MIN_PEAKLENGTH: MACS2 minimum peak length. Default (empty string) uses predicted fragment size determined by Macs2. Could be increased if broad option is used. For ATAC-Seq it is recommended to use a smaller filter threshold like 100 bp.
  - ESSENTIAL_DIFFBIND_LIBRARY: DiffBind method to calculate library sizes. One of "full", "RiP", "background" and "default" ("default" refers to method "full", see [DiffBind documentation](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf) for explanation). 
  - ESSENTIAL_DIFFBIND_NORM: DiffBind method to calculate normalization factors. One of "lib", "RLE", "TMM", "native" and "default" ("default" refers to method "lib", see [DiffBind documentation](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf) for explanation).
  - ESSENTIAL_SUMMITS: Re-center peaks in DiffBind around consensus summit (point of greatest read overlap) with peak width 2x ESSENTIAL_SUMMITS in order to provide more standardized peak intervals (0 means no re-centering, default 200). This can prevent the widening effect of peaks occurring when neighboring peaks are merged for the consensus peak set. For broader peaks this should be set to 0, which means peak sets will not be modified.  

- additional (more specialized) parameter can be given in the header files of the individual pipeline modules (see module header files linked in the flowchart for default parameter).

If differential binding analysis is selected it is required additionally:

- contrasts_diffbind.txt: tab-separated txt-file giving intended group comparisons for differential binding analysis (1 contrast per line). 
  - contrast.name: name of contrast to be used in report.
  - contrast: definition of contrast using the format '(group1-group2)' with 'group1' and 'group2' being groups defined in the 'group' column of the targets.txt file (pairwise comparisons only).
  - mmatrix: design formula to be applied for this contrast. If no further factors to be included use '~group'. A multifactor formula must be composed from the following allowable factors: Tissue, Factor, Treatment, Replicate, Caller and group (aka Condition). For diffbind version 2 multifactor design is not supported and this column is ignored (group column is used automatically).
  - sub_experiment: any name to summarize the contrasts into sub-experiments. All sample groups involved in contrasts with the same sub_experiment name are loaded into a diffbind object and are analyzed together. This includes generation of a consensus peakset and normalization of this sample set. Peaks called in samples belonging to other sub-experiments are not included in the consensus peakset. 


## Programs required
- Bedtools
- Bowtie2
- Bamutil
- Cutadapt
- deepTools
- encodeChIPqc (provided by another project from imbforge)
- FastQC
- MACS2
- MultiQC
- Picard
- R with packages ChIPSeeker, DiffBind, GenomicAlignments, GreyListChIP, spp and genome annotation packages
- Samtools
- UCSC utilities
